gflownet/gflownet.py

```diff
@@ -591,43 +643,143 @@ def trajectorybalance_loss(self, it, batch):
-        actions = batch.get_actions()
-        parents = batch.get_parents(policy=True)
-        traj_indices = batch.get_trajectory_indices()
-        masks_f = batch.get_masks_forward(of_parents=True)
-        masks_b = batch.get_masks_backward()
+        # Get logprobs of forward and backward transitions
+        logprobs_f = self.compute_logprobs_trajectories(batch, backward=False)
+        logprobs_b = self.compute_logprobs_trajectories(batch, backward=True)
+        # Get rewards from batch
         rewards = batch.get_terminating_rewards(sort_by="trajectory")
 
-        # Forward trajectories
-        policy_output_f = self.forward_policy(parents)
-        logprobs_f = self.env.get_logprobs(
-            policy_output_f, True, actions, states, masks_f
+        # Trajectory balance loss
+        loss = (
+            (self.logZ.sum() + logprobs_f - logprobs_b - torch.log(rewards))
+            .pow(2)
+            .mean()
         )
-        sumlogprobs_f = torch.zeros(
-            batch.get_n_trajectories(),
+        return loss, loss, loss
+
+    @torch.no_grad()
+    def estimate_logprobs_data(
+        self,
+        data: Union[List, str],
+        n_trajectories: int = 1,
+        max_iters_per_traj: int = 10,
+    ):
+        """
+        Estimates the probability of sampling with current GFlowNet policy
+        (self.forward_policy) the objects in a data set given by the argument data. The
+        (log) probabilities are estimated by sampling a number of backward trajectories
+        (n_trajectories) through importance sampling and calculating the forward
+        probabilities of the trajectories.
+
+        $\log p_T(x) = \int_{x \in \tau} P_F(\tau)d\tau$
+        $= \log \mathbb{E}_{P_B(\tau|x)} \frac{P_F(x)}{P_B(\tau|x)}$
+        $\approx \log \frac{1}{N} \sum_{i=1}^{N} \frac{P_F(x_i)}{P_B(\tau|x_i)}$
+        $= \log \sum_{i=1}^{N} \frac{P_F(x_i)}{P_B(\tau|x_i)} - \log N$
+
+        Note: torch.logsumexp is used to compute the log of the sum, in order to have
+        numerical stability, since we have the log PF and log PB, instead of directly
+        PF and PB.
+
+        Args
+        ----
+        data : list or string
+            A data set of terminating states. The data set may be passed directly as a
+            list of states, or it may be a string defining the path to a pickled data
+            set where the terminating states are stored in key "x".
+
+        n_trajectories : int
+            The number of trajectories per object to sample for estimating the log
+            probabilities.
+
+        max_iters_per_traj : int
+            The maximum number of attempts to sample a distinct trajectory, to avoid
+            getting trapped in an infinite loop.
+
+        Returns
+        -------
+        logprobs_estimates: torch.tensor
+            The logarithm of the average ratio PF/PB over n trajectories sampled for
+            each data point.
+        """
+        max_data = 1e5
```

Comment:
1) that's pretty hacky,
2) if it is necessary, maybe better make it an argument?

Reply:
(2. in huge part because that would force you to describe what it is and why it's necessary in a docstring)

Reply:
I think it's important to have to avoid someone passing a massive data set of say 10 M data points for which you have to sample 10 trajectories for each point...

Reply:
That's why I didn't make it an argument, but I will.